csharp/CompoundLogicalColumnReader.cs

```diff
@@ -0,0 +1,313 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using ParquetSharp.Schema;
+
+namespace ParquetSharp
+{
+    internal sealed class CompoundLogicalColumnReader<TPhysical, TLogical, TElement> : LogicalColumnReader<TElement>
+        where TPhysical : unmanaged
+    {
+        internal CompoundLogicalColumnReader(ColumnReader columnReader, int bufferLength)
+            : base(columnReader, bufferLength)
+        {
+            var converterFactory = columnReader.LogicalReadConverterFactory;
+
+            if (RepLevelsRequired(typeof(TElement)) && RepLevels == null)
+            {
+                throw new Exception("RepLevels are required but missing");
+            }
+
+            _bufferedReader = new BufferedReader<TPhysical>(Source, (TPhysical[]) Buffer, DefLevels, RepLevels);
+            _converter = (LogicalRead<TLogical, TPhysical>.Converter) converterFactory
+                .GetConverter<TLogical, TPhysical>(ColumnDescriptor, columnReader.ColumnChunkMetaData);
+
+            _reader = MakeReader(GetSchemaNode(ColumnDescriptor.SchemaNode).ToArray(), typeof(TElement), 0, 0, false);
+        }
+
+        public override int ReadBatch(Span<TElement> destination)
+        {
+            var result = (Span<TElement>) (TElement[]) _reader(destination.Length);
+
+            result.CopyTo(destination);
+
+            return result.Length;
+        }
+
+        private Func<int, Array> MakeReader(Node[] schemaNodes, Type elementType, int repetitionLevel, int nullDefinitionLevel, bool wantSingleItem)
```

Comment:
As mentioned on slack, switching to using a `defintionLevel` instead of `nullDefinitionLevel` as was done on master would probably simplify some things and seems more logical to me.